File: Bit Manipulation/getSum371.java
class Solution {
    // 011 3
    //+101 5
    //1000 8
    public int getSum(int a, int b) {
        while (b != 0) {
            int carry = a & b; // 101 & 011 = 001
            a = a ^ b; // a = 101 ^ 011 = 110
            b = carry << 1; // b = 00 << 1 = 0
        }
        return a; // a = 11 = 3
    }
    
    // recursive
    public int getSum(int a, int b) {
        return b == 0 ? a : getSum(a ^ b, (a & b) << 1);
    }
}

File: Bit Manipulation/getSum371.py
class Solution:
    def getSum(self, a: int, b: int) -> int:
        # Iteration
        # bit_digits = math.floor(math.log2(max(abs(a), abs(b)))) + 2
        # mask = 2 ** (bit_digits + 1) - 1
        # b is carrier, a stores the result
        mask = 0xffffffff
        while b:
            if b & mask == 0:
                return (a & mask)                     
            c = a & b
            a ^= b
            b = c << 1
        return a
    
        # recursion
        # a ^ b: sum, (a & b) << 1 carrier
        mask = 0xffffffff
        if b == 0: 
            return a
        if b & mask == 0: # prevent overflow
            return (a & mask)
        return self.getSum(a^b, (a&b)<<1)